python/ql/test/library-tests/frameworks/lxml/taint_test.py

```diff
@@ -0,0 +1,154 @@
+        elem.keys(),  # $ tainted
+        elem.values(),  # $ tainted
+        elem.xpath("b")[0].text,  # $ tainted getXPath="b"
+    )
+
+    for ch in elem:
+        ensure_tainted(
+            ch,  # $ tainted
+            ch.text  # $ MISSING: tainted # API node getASubscript() appears to not consider things like for loops 
+        )
+
+    buf = io.StringIO(src)
+    tree = ET.parse(buf) # $ decodeFormat=XML decodeInput=buf xmlVuln='XXE' decodeOutput=ET.parse(..) SPURIOUS:getAPathArgument=buf # Spurious as this is used as a file-like object, not a path 
+    ensure_tainted(
+        tree,  # $ tainted
+        tree.getroot().text,  # $ tainted
+        tree.find("b").text,  # $ tainted
+        tree.findall("b")[0].text,  # $ tainted
+        tree.getiterator(),  # $ tainted
+        next(tree.getiterator()).text,  # $ MISSING:tainted
+        tree.iter(),  # $ tainted
+        next(tree.iter()).text,  # $ MISSING:tainted
+        tree.iterfind("b"),  # $ tainted
+        next(tree.iterfind("b")).text,  # $ MISSING:tainted
+        tree.xpath("b")[0].text,  # $ tainted getXPath="b"
+    )
+
+    (elem2, ids) = ET.XMLID(src) # $ decodeFormat=XML decodeInput=src xmlVuln='XXE' decodeOutput=ET.XMLID(..)
+    (elem3, ids2) = ET.XMLDTDID(src) # $ decodeFormat=XML decodeInput=src xmlVuln='XXE' decodeOutput=ET.XMLDTDID(..)
+
+    ensure_tainted(
+        elem2,  # $ tainted
+        elem3,  # $ tainted
+        ids,  # $ tainted
+        ids2,  # $ tainted
+        elem2.text,  # $ MISSING:tainted # Type is not tracked to the tuple return value
+        elem3.text,  # $ MISSING:tainted
+    )
+
+    buf = io.StringIO(src)
+    (tree2,ids3) = ET.parseid(buf)  # $ decodeFormat=XML decodeInput=buf xmlVuln='XXE' decodeOutput=ET.parseid(..) SPURIOUS:getAPathArgument=buf
+
+    ensure_tainted(
+        tree2, # $ tainted
+        ids3, # $ tainted
+        tree2.getroot() # $ MISSING:tainted
+    )
+
+    buf = io.BytesIO(bytes(src, "utf8"))
+    for ev, el in ET.iterparse(buf):  # $ decodeFormat=XML decodeInput=buf xmlVuln='XXE' decodeOutput=ET.iterparse(..) SPURIOUS:getAPathArgument=buf
+        ensure_tainted(
+            el, # $ tainted
+            el.text, # $ MISSING:tainted
+        )
+
+    def func(tree_arg: ET.ElementTree):
+        ensure_tainted(
+            tree_arg, # $ tainted
+            tree_arg.getroot().text # $ tainted # Type tracking from the type hint
+        )
+
+    func(tree2)
+
+    def func2(x):
+        return x 
+
+    def func3(x) -> ET.Element:
+        return x
+
+    ensure_tainted(
+        func2(tree), # $ tainted
+        func2(tree).text, # $ MISSING:tainted - type tracking not tracked through flow preserving calls
+        func3(tree).text, # $ MISSING:tainted - this includes if there is a type hint annotation on the return
+        typing.cast(ET.ElementTree, tree), # $ tainted
+        typing.cast(ET.ElementTree, tree).text, # $ MISSING:tainted - this includes for flow summary models
```

Comment:
It will only work for "simple" summaries; those that do not use type tracking to identify calls. (And, again, is `.text` the right test here? I do not see `tree.text` being tainted.)